net: increased support for iface agnosticism

glarsennordic opened the pull request:
This is a draft/hypothetical/possibly-review-only PR related to [Changes to network stack to facilitate iface-agnostic application code](https://github.com/zephyrproject-rtos/zephyr/issues/52695)

Edit:

The PR is now a full implementation

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented:
Not yet finished or ready for review

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -4365,6 +4433,169 @@ bool net_if_is_promisc(struct net_if *iface)
+		goto out;
+	}
+
+	/* Question for reviewers:
```
In my opinion, global options should be handled by `net_if` itself. Since the current proposal makes a clear distinction between global/non-global options (with `NET_IF_GLOB_OPT_BASE`) it could be easily determined what should be forwarded down the inteface-specific implementation, and what not.

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -4365,6 +4433,169 @@ bool net_if_is_promisc(struct net_if *iface)
+	 * "net_if_api->handle_timeout" on the iface.
+	 *
+	 * This runs the risk of dipping too deeply into the inner-workings of network ifaces,
+	 * though. I lean more towards letting each iface implement its timeout mechanism how
+	 * it sees fit.
```
True, I think specific connection handlers could implement the timeout handling in their own way. Still, this has nothing to do with the `set_opt()` functionality, as this only specified the timeout value.

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -4365,6 +4433,169 @@ bool net_if_is_promisc(struct net_if *iface)
+
+/* Note to reviewers:
+ * These helper functions were not discussed in the RFC.
+ * After writing an implementation of conn_mgr, I thought some helper functions for the
```
I have no problem with having helper functinos to set/read those parameter. The question rises though, if we have such helpers, do we need to pipe them through the `net_if_*_opt()`. Provided the parameters are handled globally by `net_if`, they could be read/set directly.

There's also a question on how to store those global parameters. Given that not all interfaces will implement connectivity API, it may not be desirable to store them directly within the `net_if` structure. Perhaps the already mentioned `struct net_l2_connectivity` could come in handy, allowing to store parameters and a pointer to the connectivity API.

glarsennordic replied:
I have to agree -- persistence, timeout, max_tries seem like they would be cleaner just be directly implemented on the iface, not accessed via `net_if_*_opt`

If we do that, `net_if_*_opt` would truly be exclusively for iface-specific options, and that seems cleaner and less confusing to me, than trying to mix in these global options with it

rlubos commented on subsys/net/lib/conn_mgr/conn_mgr.c:
```diff
@@ -191,6 +191,68 @@ K_THREAD_DEFINE(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_STACK_SIZE,
+ * In light of comments made on the original RFC, rather than implement these as net_mgmt requests,
+ * I have implemented them directly as net_conn_mgr functions.
+ *
+ * Does this seem appropriate? Would net_mgmt requests be better, despite the fact it would
```
IMO the current proposal looks pretty neat.

glarsennordic replied:
@rlubos  Just to be clear, do you mean the RFC issue, or what is in this PR?

rlubos commented on subsys/net/lib/conn_mgr/events_handler.c:
```diff
@@ -39,10 +39,26 @@ static void conn_mgr_iface_events_handler(struct net_mgmt_event_callback *cb,
 	case NET_EVENT_IF_CMD_DOWN:
 		iface_states[idx] &= ~NET_STATE_IFACE_UP;
+
+		if (IS_ENABLED(NET_CONNECTION_MANAGER_AUTO_IF_DOWN) &&
```
You need to prefix Kconfig entries with `CONFIG_` when used in C, i. e. `CONFIG_NET_CONNECTION_MANAGER_AUTO_IF_DOWN`.

glarsennordic replied:
🤦My bad -- good catch

wbober commented on include/zephyr/net/net_if.h:
```diff
@@ -2163,6 +2169,25 @@ static inline bool net_if_is_up(struct net_if *iface)
+ *
+ * @param iface Pointer to network interface
+ *
+ * @return 0 on success, -ESHUTDOWN if the iface is not admin-up, implementation-specific status
```
I believe that we also have `ENOTSUPP` flag defined. Could be also included as an expected result code.

wbober commented on include/zephyr/net/net_if.h:
```diff
@@ -2163,6 +2169,25 @@ static inline bool net_if_is_up(struct net_if *iface)
+ *
+ * @param iface Pointer to network interface
+ *
+ * @return 0 on success
```
ditto

wbober commented on include/zephyr/net/net_if.h:
```diff
@@ -2357,6 +2382,113 @@ void net_if_unset_promisc(struct net_if *iface);
+ * @return True if the option is enabled, otherwise False.
+ * Also returns false if the persistence value could not be retrieved or is not supported.
+ */
+bool net_if_is_connectivity_persistent(struct net_if *iface);
```
Subjection but personally, I'd just drop these wrapper functions for the sake of API simplicity. If we see that some options are used very often then we can introduce them.

glarsennordic replied:
I think I might agree if not for the fact that @rlubos has made a pretty convincing argument that maybe these options shouldn't be part of `net_if_get_opt` at all, in which case they cease to be helpers (and also become much simpler) -- What do you think of going in that direction instead?

wbober commented on subsys/net/ip/net_if.c:
```diff
@@ -4196,6 +4205,65 @@ int net_if_up(struct net_if *iface)
+
+	const struct net_if_api *api = net_if_get_device(iface)->api;
+
+	if (api->connect) {
```
If `api->connect` is null should the status be still 0?

glarsennordic replied:
I go back and forth on whether these functions should yield ENOTSUP, or just no-op for ifaces that don't have a concept of connectivity.

Probably better to return ENOTSUP, and then have `conn_mgr` intelligently handle the error, or otherwise detect whether the iface supports connectivity

wbober commented on subsys/net/ip/net_if.c:
```diff
@@ -4196,6 +4205,65 @@ int net_if_up(struct net_if *iface)
+
+	const struct net_if_api *api = net_if_get_device(iface)->api;
+
+	if (api->disconnect) {
```
ditto

rlubos commented on subsys/net/lib/conn_mgr/events_handler.c:
```diff
@@ -39,10 +39,26 @@ static void conn_mgr_iface_events_handler(struct net_mgmt_event_callback *cb,
+		if (IS_ENABLED(NET_CONNECTION_MANAGER_AUTO_IF_DOWN) &&
+		    !net_if_is_connectivity_persistent(iface)) {
+			/*
+			 * Take the iface admin-down, if it isn't already.
```
I think it should be fine to bring the interface admin down directly from the cb.

glarsennordic replied:
I thought I had read somewhere that firing a `net_mgmt` event from within the callback of another is a bad idea, but I can't find where I read that, and I'm starting to think I am conflating. It sounds like that's a non-issue, then?

glarsennordic replied:
Working on the L2 dummy, I stumbled across where I read that warning. It's in `ethernet.h`:

> Carrier ON/OFF handler worker. This is used to create
> network interface UP/DOWN event when ethernet L2 driver
> notices carrier ON/OFF situation. We must not create another
> network management event from inside management handler thus
> we use worker thread to trigger the UP/DOWN event.

It's not actually clear to me **_why_** this is the case, but I'll dig into it.

jukkar replied:
Main reason for this is that the net_mgmt thread stack is quite small and will overflow if we try to do another net mgmt event + other operations from inside the same thread.

glarsennordic replied:
Thanks for the explanation, @jukkar -- I have delegated the net_if_down call to the conn_mgr thread accordingly

rlubos reviewed (commented):
Left some initial feedback, mostly for further discussion. Thanks for the effort @glarsennordic!

wbober commented on subsys/net/ip/net_if.c:
```diff
@@ -4365,6 +4433,169 @@ bool net_if_is_promisc(struct net_if *iface)
+	 *
+	 * I am not sure whether it would be cleaner to:
+	 *
+	 * 1) Let each iface that supports either NET_IF_OPT_CONNECTIVITY_PERSISTENT or
```
I think I'd support that for now.

wbober commented on subsys/net/ip/net_if.c:
```diff
@@ -4365,6 +4433,169 @@ bool net_if_is_promisc(struct net_if *iface)
+ * After writing an implementation of conn_mgr, I thought some helper functions for the
+ * global options would be nice.
+ *
+ * What do you think?
```
I'd keep the net_if API small. You can move these as statics to `conn_mgr` for now.

wbober commented on subsys/net/lib/conn_mgr/conn_mgr.c:
```diff
@@ -191,6 +191,68 @@ K_THREAD_DEFINE(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_STACK_SIZE,
+ * In light of comments made on the original RFC, rather than implement these as net_mgmt requests,
+ * I have implemented them directly as net_conn_mgr functions.
+ *
+ * Does this seem appropriate? Would net_mgmt requests be better, despite the fact it would
```
Looks good to me.

glarsennordic commented on include/zephyr/net/net_event.h:
```diff
@@ -53,6 +54,20 @@ enum net_event_if_cmd {
 	(_NET_EVENT_IF_BASE | NET_EVENT_IF_CMD_ADMIN_UP)
 
+/* Note to reviewers:
+ * This likely should be renamed if we keep NET_IF_OPT_CONNECTIVITY_MAX_TRIES
```
@rlubos 

To me, persistent connectivity is connectivity that, at least with default settings, never gives up on being connected.

To be clear, `NET_IF_OPT_CONNECTIVITY_MAX_TRIES`, is only a consecutive counter. If connection succeeds, the total number of tries will be reset to zero. Some ifaces may not even have a concept of separate connection events, and on such ifaces `NET_IF_OPT_CONNECTIVITY_MAX_TRIES` could be ignored.

So, with an iface that is persistent, the  `NET_IF_OPT_CONNECTIVITY_MAX_TRIES` and `NET_IF_OPT_CONNECTIVITY_TIMEOUT` options basically specify special conditions under which even a **_persistent_** connection should give up (specifically when trying to transition from dormant to oper-up)

That means that:
> It's up to application to stop the cycle if persistence is enabled

Is only true if `NET_IF_OPT_CONNECTIVITY_MAX_TRIES` and `NET_IF_OPT_CONNECTIVITY_TIMEOUT` are both zero, at least as the RFC is currently written.

Having this built in **_may_** be a good idea just because it is a very common desire among actual applications -- But it also wouldn't be the end of the world if applications have to implement these behaviors themselves using events (in which case we should add an event specifically for "a connection attempt failed, but I am still trying to connect"). So really, it would be good to get input from others.

But that's just for persistent connectivity. What about non-persistent connectivity?

Conceivably, non-persistent connectivity could be defined two ways:

- Will try repeatedly to connect on the ***first*** transition from dormant to oper-up, but if connection is ever lost after that, gives up
- Will try only once, ever, to connect

Maybe the second definition is cleaner to adopt, since it allows the `MAX_TRIES` option to be for persistent networks only.

But even adopting that second definition, `CONNECTIVITY_TIMEOUT` still applies, and if it is infinite, even a non-persistent iface could wind up trying to connect forever anyways -- in which case, maybe `MAX_TRIES` ***should*** be an optional parameter for non-persistent ifaces.


I am neutral on both decisions. We just need to make a decision. To summarize:

1. Decide whether connection timeout and connection max_tries should be built-in, or up to the application to implement
2. If built-in, decide whether max_tries should be defined for non-persistent ifaces as well (and apply only during the first attempt to transition from dormant to oper-up)

wbober replied:
I think this feature is tricky because underlying protocols might differ how they approach the issue. I'd be even keen to park the persistence feature to gather more direct feedback about L2s more rather than make an API that isn't well aligned.

Having said that, I personally think that in this case it's a bit easier to deal with time than with retries. This is for two reasons:

1. In order to implement retries the underlying L2 must report connection retry attempts. This might not be available.
2. It might be easier to reason about power consumption. I could assume the worst case power consumption for connection attempts (CPU on, Radio On) and calculate energy expenditure. With retries I need to understand particular L2 behavior. 

I would prefer not to distinguish between first and consecutive connection attempts to make the API easier to grasp.

How about simplifying as follows:

* Keep `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` opt and remove everything else.
* For the `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` set to a non zero value it is expected that L2 will attempt to change from `dormant(true)` to `dormant(false)` state for at most `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` seconds.
* For `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` is set to 0 (?) the behavior is L2 implementation specific.

glarsennordic replied:
You're right, introducing max_tries does require us to distinguish between two different types of connection attempt: The "macro" attempt, which may (or may not, depending on the L2) be a series of "micro" attempts. That adds confusion.

I'm on-board with the simplification you propose. It has the added benefit of not distinguishing between first and subsequent "macro" attempts "and also it doesn't distinguish between micro and macro connections at all", and I think that simplification is worth the minor loss of functionality.

glarsennordic commented on include/zephyr/net/net_if.h:
```diff
@@ -2357,6 +2382,113 @@ void net_if_unset_promisc(struct net_if *iface);
+ *
+ * @param optlen Pointer to length (in bytes) of the destination buffer available for storing the
+ *		 retrieved value. If the available space is less than what is needed, or otherwise
+ *		 invalid, it is up to the interface implementation whether to:
+ *		 1) Truncate the retrieved value and return -ENOBUFS
+ *		 2) Refuse (with an error code) to retrieve the value at all
```
Interestingly, for some types (like int) some getsockopt options do follow the second behavior, despite the official `getsockopt` description claiming they do not.


Third option is:
- If the buffer provided is not large enough to fit the option, return an error. If the buffer was written to in the process of determining that, the size reference will be set to the number of bytes written

I think I will adopt that approach if there aren't any objections

glarsennordic commented on subsys/net/ip/net_if.c:
```diff
@@ -4047,6 +4047,15 @@ static void notify_iface_up(struct net_if *iface)
 	}
+
+	/* Question for reviewers:
+	 * Is this the best place to do this? Do you know/recommend another spot?
```
Ahhh, you're right, good catch, thank you. Even now I can't seem to stop myself from thinking of iface_up as admin_up, not oper_up. I'll get it eventually.

glarsennordic commented on subsys/net/ip/net_if.c:
```diff
@@ -4196,6 +4205,65 @@ int net_if_up(struct net_if *iface)
+	}
+
+	/* Question to reviewers:
+	 * Would it be cleaner to attach the underlying API implementation to the
```
I really like the idea of a separated net_l2_connectivity API, not only for the reasons you list, but also since it creates a true structural difference between ifaces that have a concept of connectivity, and those that don't. I'll give implementing that a shot next week and see how we like it, unless @jukkar or @wbober are not so convinced

As for the macrobatics, if we make it an optional param, would it be better to define a few alternate _CONNECTIVITY macros that take the extra argument, or really go for a `__VA_ARGS__` approach?

wbober replied:
I think `struct net_l2_connectivity` is an elegant solution.

glarsennordic commented on subsys/net/ip/net_if.c:
```diff
@@ -4196,6 +4205,65 @@ int net_if_up(struct net_if *iface)
+	/* Question to reviewers:
+	 * Would it be cleaner to attach the underlying API implementation to the
+	 * net_l2 struct instead...? This would require modifications be made to allow
+	 * offloaded sockets to have L2 implementations. But I think we will need this
```
I agree, let's do both the offloaded net_l2 dummy, and your `net_l2_connectivity` suggestion. I guess this would be the right PR to get started on the dummy, since we know at this stage we'll want it down the line.

plskeggs commented on include/zephyr/net/net_if.h:
```diff
@@ -2357,6 +2386,114 @@ void net_if_unset_promisc(struct net_if *iface);
+ *
+ * @retval 0 if successful
+ * @retval -ENOTSUP if net_if_set_opt is not implemented by the iface
+ * @retval -ENOBUFS if not enough retrieval buffer is provided
```
```suggestion
 * @retval -ENOBUFS if retrieval buffer is too small
```

rlubos commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,120 @@
+ */
+struct net_if_conn {
+	/* The network device the connectivity struct is bound to */
+	struct net_if_dev *if_dev;
```
Orignally I though it'd be implemented the other way around (`net_if_dev` keeping a pointer to `net_if_conn`) but this could also work. Maybe it's even nicer, as there is no RAM penalty (pointer) for interfaces that do not use L2 connectivity.

glarsennordic replied:
That is what I had been thinking too (and what I had implemented until recently), but the issue I encountered is that it becomes impossible to have a separate macro for registering the connectivity API since C doesn't allow structs to be modified at build time, and neither does it offer any mechanism to create mixin initializers.

So if we went the other route (net_if_dev keeping a pointer to net_if_conn), then we could not have the separated NET_DEVICE_DEFINE_L2_CONNECTIVITY macro, and net_if_conn would instead have to be an additional argument to NET_DEVICE_INIT and its ilk.

Going via the section lists is pretty much the only work-around to this issue I could find that operates at compile-time, but it is possible I'm just missing some other technique that could accomplish it! Do you know of any?

I'm fine going with whichever solution folks here think would be cleanest / most canonical. This was just my attempt to avoid further complicating the NET_DEVICE_INIT argument list

rlubos replied:
I'm personally fine with the current approach. I'd be more concerned if the connectivity API was heavily used due to additional overhead for finding the corresponding `net_if_conn` structure, but it's not the case here.

rlubos commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,120 @@
+	int (*disconnect)(struct net_if *iface);
+
+	/* Note to reviewers:
+	 * Technically, as long as each iface provides a means by which to set
+	 * association/connection parameters, these get/set_opt funcs are not needed.
+	 *
+	 * However, it can be argued that they should still exist, since they establish a
+	 * standardized pattern for doing that.
```
Wasn't the whole point of those functions to provide a generalized way to configure connection parametrs? IMO they should stay.

rlubos commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,120 @@
+	 * However, it can be argued that they should still exist, since they establish a
+	 * standardized pattern for doing that.
+	 *
+	 * The question, then, is: Why limit these to specifically connectivity?
```
Personally, I'd stick to connectivity only. Those sort of functions are already pretty obfuscated by their nature, if they get too much responsibility (i. e. configuring every interface param one could imagine) it could get quite confusing.

wbober replied:
Plus, if we don't have any particular use case on the horizon that would justify this we should not add it.

https://en.wikipedia.org/wiki/You_aren't_gonna_need_it

glarsennordic replied:
I'm somewhat relieved you both feel this way, after having sat with the question for a few days myself, this is where my opinion landed too

rlubos commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,120 @@
+	int timeout;
+	bool persistence;
+
+	/* To reviewers: Perhaps connectivity settings should be per-l2, rather than per-iface?
```
I think we're missing in this picture some kind of connectivity-implementation-specifc context structure, similar to `l2_data` concept for L2. `net_if_conn` could then keep pointers to API and data structures, plus generic parameters, like those above.

wbober replied:
Do you foresee much headache to add it if we need it?

rlubos replied:
Probably not much of an headache, but I kind of miss the point of having `set_opt()`/`get_opt()` w/o some per-interface data container.

glarsennordic replied:
I lean towards agreeing with @rlubos on this -- For our purposes, we don't foresee any ifaces sharing connectivity implementations, but since that is something we technically support with this approach, we should probably offer context registration

I'll check out how much work it would be to implement, and if I think I can get it in today, I'll do it. Otherwise, I'll leave it for when we review with a wider audience

glarsennordic replied:
**Firstly:**

So I am realizing this whole thing is a bit of a miscommunication on my part.

I called `timeout` and `persistence` "connectivity settings", which I meant to be distinct from the settings controlled by `set_opt` and `get_opt`.

So due to my miscommunication, the question I originally intended to ask still stands:

Should timeout, persistence be per-iface, or per-L2?


**Secondly:**

2.) @rlubos You mention a connectivity-implementation-specifc context structure, but then later mention `per-interface data container.`

I think that per-connectivity-implementation data could be stored statically along-side the implementation, so I am not certain there is a need for it.

The set_opt and get_opt functions, would then serve as generic accessors to whatever state the implementation has built for itself, so they would still serve purpose.


**Thirdly:**
I do think you are right, @rlubos -- as-written, there is no way for a connectivity implementation to store per ***netdev/iface*** data.

For this purpose, we should probably provide some kind of context pointer, although I am still trying to work out what I think the best form for it would be.

rlubos replied:
> But storing it per-iface also seems a bit odd, since several ifaces might share the same L2 connection (suppose there are several ifaces assigned to LTE for instance).

IMO **all** connection parameters should be stored at per interface basis. Imagine two Wi-Fi devices, exposing two network interfaces. They may share the connectivity implmentation, but they should have a chance to connect to different networks with different parameters. 

I'm not sure I follow your LTE example. All _real_ network interfaces should have a network device underneath. If we want to have multiple network interfaces working on top of a single device, then we talk about virtual interfaces, but they still need a _real_ interface underneath. In this scenario, the virtual interfaces should not deal with connectivity at all, it should be the responsibility of the underlying interface.

> @rlubos You mention a connectivity-implementation-specifc context structure, but then later mention per-interface data container.

Sorry for not being clear enough. What I meant is that we should follow the L2 data model, where the L2 declares a context structure it expects for the interface to allocate. So taking Ethernet as an example, the Ethernet L2 declare the context structure format it expects to deal with:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ethernet.h#L520
But every Ethernet interface defines a unique instance of the structure:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ethernet.h#L844

So to summarize, mutliple interfaces may share the same L2 implementation (L2 API) but they still allocate unique context structures for L2 to deal with for each interface. I think it only makes sense to follow the same pattern for connectivity.

> I do think you are right, @rlubos -- as-written, there is no way for a connectivity implementation to store per netdev/iface data.
For this purpose, we should probably provide some kind of context pointer, although I am still trying to work out what I think the best form for it would be.

I'll again refer to L2 model and `NET_L2_DATA_INIT()` macro. There is some macrology behind it, but the sole purpose of this macro is to allocate a context structure, defined by L2, uniquely for the interface. The structure is then linked with the interface in `NET_IF_INIT()` using `NET_L2_GET_DATA()` to obtain the context structure pointer.

glarsennordic replied:
>  They may share the connectivity implementation, but they should have a chance to connect to different networks with different parameters.

You're right, a board designer could have, for example, two instances of the same Wi-Fi device model, so it is not accurate for me to assume that if two ifaces share an implementation, they are also sharing a physical device. 

>  All real network interfaces should have a network device underneath. If we want to have multiple network interfaces working on top of a single device, then we talk about virtual interfaces, but they still need a real interface underneath.

My bad, I was operating under the misunderstanding that non-virtual network devices could share physical devices. If this is disallowed, then storing timeout and persistence per-iface seems perfectly reasonable.


These two corrections in mind, I completely agree with your stance. Fortunately, I speculatively worked on an implementation yesterday assuming I would wind up agreeing, so most of this is already in place. I will try to have it merged some time today.

rlubos commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,120 @@
+ * @param inst Network device instance.
+ * @param conn_api Pointer to net_l2_connectivity_api struct to associate.
+ */
+#define NET_DEVICE_INSTANCE_DEFINE_L2_CONNECTIVITY(dev_id, inst, conn_api)	\
```
I wonder if we could follow the L2 approach here, i. e. instead of passing a pointer to the connectivity API struct, implement something similar to `NET_L2_INIT()` and `NET_L2_DECLARE_PUBLIC()` where specific connectivity implementation could register their existence and then be referenced here by name.

wbober replied:
If we go with this direction, i.e., `NET_L2_CONN_INIT()` and `NET_L2_DECLARE_PUBLIC()` then I would probably be more glad to see `NET_DEVICE_INIT()` macros modified to take the name of L2 CONN just as it takes L2 name. At least to me this would make it more logical compared to an init in a separate place.

While pitching this to Robert he pointed out that we should define the L2 CONN structures only if needed. I think that `COND_CODE_0` macro could help. How, I think, this could work is that if you don't use a L2 CONN then you pass a something that expands to 0 (NULL?). This in turn places empty block of code.

On the other hand if you pass a valid L2 CONN name, e.g., L2_CONN_WIFI then this would expand to `NET_DEVICE_INSTANCE_DEFINE_L2_CONNECTIVITY(..., ..., L2_CONN_WIFI)`;

glarsennordic replied:
> While pitching this to Robert he pointed out that we should define the L2 CONN structures only if needed. I think that COND_CODE_0 macro could help. How, I think, this could work is that if you don't use a L2 CONN then you pass a something that expands to 0 (NULL?) to the modified NET_DEVICE_INIT() macro. This in turn places an empty block of code.

Yeah, were we to go this route, we'd just have callers of `Z_NET_DEVICE_INIT` or any of its wrappers/equivalents pass NULL for the l2_conn argument on interfaces that don't need a connectivity implementation

But I hesitate to add an argument to `NET_DEVICE_INIT`. Firstly, there are a total of `18` associated macros in Zephyr that would also have to be modified to accept this additional argument. And, in the Zephyr codebase alone, these macros are referenced a total of `172` times in `92` files, and the majority of these references would be simply passing NULL to the new argument.

Basically, I worry that further complicating this already pretty complex call tree may have long term detrimental effects on maintainability.

I'm willing to do it if there is strong agreement that it really is the way to go, but I just want to lift up that caution before doing so

wbober replied:
Yeah, I understand. And then I don't expect the L2 connectivity implementations to be added often, so I think what you propose is fine.

rlubos commented on subsys/net/lib/conn_mgr/conn_mgr.c:
```diff
@@ -24,6 +24,16 @@ LOG_MODULE_REGISTER(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL);
 
 uint16_t iface_states[CONN_MGR_IFACE_MAX];
 
+/**
```
I have an impression that those `conn_mgr` changes are not tightly bound to the generic connectivity framework introduction. Perhaps they could be provided in a separate commit to simplify the review process?

wbober replied:
And perhaps come with a modification of a sample to demonstrate the new approach?

rlubos reviewed (commented):
Some nits + architecture-related comments. I'm glad to see it's slowly getting into final shape :+1:

PerMac commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -47,7 +47,8 @@ class SizeCalculator:
         "net_l2_area",
         "net_l2_data",
-        "k_queue_area",
+        "net_if_conn_area", # Reviewers: It is not clear to me whether I should add this entry.
```
sorry, no idea. We had some issues with this calculations and I don't know how reliable they are and how they are used. @nashif @stephanosio (I only see your commits in the history of size_calc)?

glarsennordic commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,120 @@
+	 * For now, I have implemented them just on the connectivity_api, but I wonder if they
+	 * should perhaps be more generalized.
+	 */
+	int (*set_opt)(int optname, const void *optval, size_t optlen);
```
Yes! You're absolutely right

glarsennordic committed:
net: increased support for iface agnosticism

Adds connectivity API abstraction and declarable
L2 connectivity implementation structures that
can be associated with network devices and used
to manage post-admin-up connect/disconnect

Implementation of #52695

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

rlubos reviewed (commented):
For me, the PR looks pretty good already. Thanks @glarsennordic!

glarsennordic committed:
net: dummy L2 for offloaded ifaces

Adds dummy link layer for offloaded ifaces, allowing
ifaces to directly receive l2_enable calls

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
tests: net: Add tests for offloaded_netdev

Adds integration tests for offloaded_netdev iface API extension

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented on tests/net/offloaded_netdev/src/main.c:
```diff
@@ -0,0 +1,343 @@
+	atomic_set(&call_count, 0);
+	atomic_set(&up_count, 0);
+}
+
```
Write descriptions for tests

glarsennordic commented:
I am taking this PR out of draft; **_It should still remain DNM_**, and I still need to write documentation and some additional automated tests, but at this point the core code is ready for review.

rlubos commented:
Keeping `DNM` for now as more tests/docs are pending, but the core functionality is ready for more reviews.

glarsennordic committed:
tests: net: Add tests for l2_connectivity

Adds integration tests for l2_connectivity features

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

simensrostad commented on include/zephyr/net/net_l2_connectivity.h:
```diff
@@ -0,0 +1,135 @@
+	void *ctx;
+
+	/* Generic connectivity state */
+	bool persistence;
```
Would be helpful with some additional docs here. Especially that the timeout is in seconds.

glarsennordic commented on tests/net/socket/offload_dispatcher/src/main.c:
```diff
@@ -8,6 +8,7 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/net/dummy.h>
 #include <zephyr/net/net_if.h>
+#include <zephyr/net/offloaded_netdev.h>
```
Change accidentally introduced in the wrong commit. Correct this on a rebase.

glarsennordic commented:
I've pushed an unfinished (and currently failing) tests commit for the sake of giving more context sooner to reviewers

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -4249,6 +4280,73 @@ int net_if_down(struct net_if *iface)
+{
+	NET_DBG("iface %p connect", iface);
+
+	const struct net_if_conn *connectivity = net_if_get_conn(iface);
```
It would be nice if we could follow zephyr style here, so variables introduced in the beginning of the block.
I see this in this function and the other functions too.

jukkar commented on subsys/net/lib/conn_mgr/Kconfig:
```diff
@@ -35,4 +35,8 @@ config NET_CONNECTION_MANAGER_PRIORITY
 	help
```
In the commit message `net: conn_mgr: Agnostic iface control multicasting`, the wording "multicast" is really confusing to me. I know IP multicasting but that is probably not the same thing? Could we change this to something less confusing.

glarsennordic replied:
I've renamed the commit message to `Bulk iface commands`

jukkar commented on subsys/net/lib/conn_mgr/conn_mgr.c:
```diff
@@ -191,6 +209,127 @@ K_THREAD_DEFINE(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_STACK_SIZE,
 		(k_thread_entry_t)conn_mgr_handler, NULL, NULL, NULL,
 		THREAD_PRIORITY, 0, 0);
 
+static void conn_mgr_all_if_up_cb(struct net_if *iface, void *user_data)
```
The function names are quite long, would it harm if we make them shorter. So something like
`conn_mgr_up_cb()` etc. The "all_if" seems redundant.

jukkar commented on subsys/net/lib/conn_mgr/conn_mgr.h:
```diff
@@ -49,4 +51,41 @@ enum net_conn_mgr_state {
+ * @retval 0 if all calls to net_if_up succeed.
+ * @retval otherwise, the error code returned by the first failed call to net_if_up.
+ */
+int net_conn_mgr_all_if_up(void);
```
This one brings all network interface up, would we need a variant where user could say that he/she wants to bring only some parts of the interface up (like some network type etc). Or a variant where it is enough to bring just one interface up. In this case we could add here some parameters to this function.

jukkar commented:
Just to clarify, to me it looks like the connection manager changes here concentrate to one use case, bringing all the network interface up/down/connect etc. IMHO this is a bit limiting and not the first use case that I would have thought.
I think we need to think more about the use case and then try to introduce the functionality to implement that.
It might be that this is already discussed in the issue but there was just too much text there so TL;DR (sorry ;)

So I am trying here to list some of the use cases I would see we need (some of these might have already be implemented):
* Application wants to have a network connection (anything goes)
* Application wants to connect to a certain type network (interface does not matter, only type)
* Application wants to connect to certain types of network in certain order
* Application wants to have all network types connected (could be done by just foreaching the other API functions)
* Application wants to disconnect an established connection
* Application wants to disconnect all the network connections
* Application wants to disconnect certain network type connection

I do not see a need for the application to know nor care for up/down stuff when thinking from connection manager point of view. I mean it should be enough for the application to just want to establish a connection or disconnect one.
If application wants to have network interfaces up or down, then it can just use lower net_if API for that.

WDYT?

@tbursztyka I know we have discussed these connection management things in the past, just wondering did I miss something above.

glarsennordic commented:
You're right, conn_mgr does focus on a pretty specific use-case, which is essentially:

- Tracking whether or not we have connectivity **_at all_**
- Controlling whether or not we have connectivity **_at all_**

Anything more precise than this starts becoming very application and iface-specific. I don't think there is even a reliable way to detect network type within Zephyr -- Wi-Fi ifaces may be offloaded, or may be based on the Ethernet L2, or possibly even both, for instance

Therefore, the conclusion that I have been forced to come to is that anything more specific than broad-spectrum control really has to be accomplished by using the single-iface `net_if_connect` and `net_if_disconnect` calls, rather than `conn_mgr_all_if_connect` and `conn_mgr_all_if_disconnect` calls.

This doesn't mean `conn_mgr_all_if_connect` and `conn_mgr_all_if_disconnect` are completely useless, though. Applications that will only ever be built for a single connectivity type at a time, for instance, are what I imagine will primarily use them.

I don't think adding filtered variants of `conn_mgr` functions that somehow target specific iface types is a bad idea, though! But I think it might be better left for a future PR, especially since right now, filtering by iface type really isn't feasible (so far as I am aware). (Maybe we need to introduce some special iface **_tags_** that iface declarations can use to mark what type they are? Maybe even bit flags?)

I am also really hesitant to implement anything like iface prioritization lists, since situations where iface prioritization / connection ordering is desired are likely to be **_extremely_** idiosyncratic. I'm not convinced that a satisfying unified implementation exists for that use case. But if it does, that's another feature I'd say is better left to a future PR

Right now, the main point is to add parity to the existing `conn_mgr` functionality (currently, `conn_mgr` can **_detect_** general iface state, and we are adding the ability to, very loosely, control general iface state)

jukkar commented:
> I don't think there is even a reliable way to detect network type within Zephyr -- Wi-Fi ifaces may be offloaded, or may be based on the Ethernet L2, or possibly even both, for instance

Hmm, I would say we have a very reliable way to say what "type" i.e., what L2 is being used by a network interface is.
So it might be that we are talking about different things here.

glarsennordic commented:
I'm not sure the L2 would be a very useful filter, since the majority of our ifaces either use the Ethernet L2, or are offloaded (and, until this PR, don't have an L2, and after this PR wind up using the offloaded iface L2)

glarsennordic committed:
net: conn_mgr: rename flags and events

Use less ambiguous and more consistent names for conn_mgr flags
and events.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Clean up state change tracking

Refactor conn_mgr_act_on_changes and associated
functions to be easier to read and maintain

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

krish2718 commented on subsys/net/lib/conn_mgr/conn_mgr.c:
```diff
@@ -191,6 +209,127 @@ K_THREAD_DEFINE(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_STACK_SIZE,
+		return;
+	}
+
+	status = net_if_down(iface);
```
This should be `net_if_connect`

glarsennordic replied:
Yikes, good catch -- I'm glad unit tests for conn_mgr are the next thing on my agenda

glarsennordic committed:
net: conn_mgr: Bulk iface commands and connectivity control

Adds commands allowing bulk control of ifaces admin and oper state
without knowledge of their L2 specifics. Also adds some automatic
behaviors related to connectivity control

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Add thread-safety

Current implementation does not prevent the net_mgmt thread from
preempting during state change handling and generating more
state changes. Add a mutex to protect internal state of conn_mgr

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Add connectivity counting

conn_mgr now tracks the total number of L4-ready
ifaces, and sends special events when this count
transitions to and from zero.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

tbursztyka reviewed (changes requested):
This seems like the wrong approach. It's mixing up concepts that should not be part of l2/iface.

iface/l2 and connectivity... no.
an interface can be up and running, on an l2 level it can be associated... connected: never. That's an IP layer concept.

zephyr net stack is basically layered as:
`device - net_if- ip stack - applications`
`            |`
`            l2`

And this PR is messing up things. (not to mention that it is bloating things up as well).
You do not connect/disconnect an interface through the l2 layer, and thus all these structures are miss-placed.

It seems to me you are trying to add features that would solve things up for what you are trying to do with conn_mgr. That's a wrong approach.
If you want to improve conn_mgr, do it there only. If you miss atomic tweaking functions for net_if, ok that can be added for net_if.

There are different types of interfaces (wifi, ethernet, etc...), it is not up to the net stack to hide the specific logic of each one for the sole purpose of a service above that (i.e. conn_mgr in this case).

rlubos commented:
> This seems like the wrong approach. It's mixing up concepts that should not be part of l2/iface.
> 
> iface/l2 and connectivity... no. an interface can be up and running, on an l2 level it can be associated... connected: never. That's an IP layer concept.

I wonder if it's just ambiguity about the naming? Putting conn_mgr aside, what we propose in https://github.com/zephyrproject-rtos/zephyr/pull/53118/commits/2335e6ba42762183023a1c1146336e75d4470eaf is a set of generic routines for interfaces to connect/associate with the network, before even IP gets involved. For example, Thread inteface needs to join Thread network  before it can be considered operational, Wi-Fi interface needs to connect to an AP. The conclusion we had internally, and which was discussed during the networking forum, was that this kind of association with the network could be considered as an L2 concept. 

> And this PR is messing up things. (not to mention that it is bloating things up as well).

Can you elaborate on what's actually getting bloated? The L2 connectivity API is not affecting network interface API in anyway, it can only be impelmented/used by intefaces that really need it. So in case such functionality is not need by a particular interface, there's no overhead at all, no single field was added to the `struct net_if` and related.

> You do not connect/disconnect an interface through the l2 layer, and thus all these structures are miss-placed.

I think I need to understand first, why exactly do you mean by connecting an interface. Our definition/understanding was as explained above, joining/associating with the network.

> It seems to me you are trying to add features that would solve things up for what you are trying to do with conn_mgr. That's a wrong approach.

Actually, conn_mgr is just an addition here. The core issue we try to address, which has become more appratent when we started writing networking applicaitons that should semleassly work with various networking technologies (Wi-Fi/LTE/Thread), is dealing with low-level network management. So that instead of dealing with this at the application level each time, speicific interface types could provide routines to deal with L2 connectivity, exposed under a generic API.

tbursztyka commented:
> > This seems like the wrong approach. It's mixing up concepts that should not be part of l2/iface.
> > iface/l2 and connectivity... no. an interface can be up and running, on an l2 level it can be associated... connected: never. That's an IP layer concept.
> 
> I wonder if it's just ambiguity about the naming? Putting conn_mgr aside, what we propose in [2335e6b](https://github.com/zephyrproject-rtos/zephyr/commit/2335e6ba42762183023a1c1146336e75d4470eaf) is a set of generic routines for interfaces to connect/associate with the network, before even IP gets involved. For example, Thread inteface needs to join Thread network before it can be considered operational, Wi-Fi interface needs to connect to an AP. The conclusion we had internally, and which was discussed during the networking forum, was that this kind of association with the network could be considered as an L2 concept.
> 
> > And this PR is messing up things. (not to mention that it is bloating things up as well).
> 
> Can you elaborate on what's actually getting bloated? The L2 connectivity API is not affecting network interface API in anyway, it can only be impelmented/used by intefaces that really need it. So in case such functionality is not need by a particular interface, there's no overhead at all, no single field was added to the `struct net_if` and related.
> 
> > You do not connect/disconnect an interface through the l2 layer, and thus all these structures are miss-placed.
> 
> I think I need to understand first, why exactly do you mean by connecting an interface. Our definition/understanding was as explained above, joining/associating with the network.
> 
> > It seems to me you are trying to add features that would solve things up for what you are trying to do with conn_mgr. That's a wrong approach.
> 
> Actually, conn_mgr is just an addition here. The core issue we try to address, which has become more appratent when we started writing networking applicaitons that should semleassly work with various networking technologies (Wi-Fi/LTE/Thread), is dealing with low-level network management. So that instead of dealing with this at the application level each time, speicific interface types could provide routines to deal with L2 connectivity, exposed under a generic API.

tbursztyka commented:
sorry, wrong manipulation

tbursztyka commented:
> I wonder if it's just ambiguity about the naming? Putting conn_mgr aside, what we propose in [2335e6b](https://github.com/zephyrproject-rtos/zephyr/commit/2335e6ba42762183023a1c1146336e75d4470eaf) is a set of generic routines for interfaces to connect/associate with the network, before even IP gets involved. For example, Thread inteface needs to join Thread network before it can be considered operational, Wi-Fi interface needs to connect to an AP. The conclusion we had internally, and which was discussed during the networking forum, was that this kind of association with the network could be considered as an L2 concept.

Thread joins, WiFi associates, ethernet does nothing, bluetooth does its own thing... 

Knowing that each tech might also require a set of specific parameters. Not to mention storage for some interface types (temporary credentials, etc...), which is all but generic.

To me you are trying to generalize concepts that cannot really be.
And so far, the current stack already provide the means for doing all of this, per-interface type basis. Which all the stack need to provide. 

And naming is extremely important, it brings semantics with it. If you put "connect" in l2 layer, so what is connect in ip layer then? It's messing up things. So then you get a connected iface, but it's not because it does not have any IP yet ? etc...

> 
> > And this PR is messing up things. (not to mention that it is bloating things up as well).
> 
> Can you elaborate on what's actually getting bloated? The L2 connectivity API is not affecting network interface API in anyway, it can only be impelmented/used by intefaces that really need it. So in case such functionality is not need by a particular interface, there's no overhead at all, no single field was added to the `struct net_if` and related.

It's adding up things in the net stack where it should no belong to.
All of this should be application layer.

> 
> > You do not connect/disconnect an interface through the l2 layer, and thus all these structures are miss-placed.
> 
> I think I need to understand first, why exactly do you mean by connecting an interface. 

The PR is adding an L2 level API about connecting/disconnecting. See struct net_l2_conn_api.
The whole idea of "l2 connectivity" is wrong by naming already.

> 
> > It seems to me you are trying to add features that would solve things up for what you are trying to do with conn_mgr. That's a wrong approach.
> 
> Actually, conn_mgr is just an addition here. The core issue we try to address, which has become more appratent when we started writing networking applicaitons that should semleassly work with various networking technologies (Wi-Fi/LTE/Thread), is dealing with low-level network management.  So that instead of dealing with this at the application level each time, speicific interface types could provide routines to deal with L2 connectivity, exposed under a generic API.

A "generic" API that will require, for each interface type, specific parameters. It's not so generic anymore: application will **anyway** need to know the interface type and behave accordingly.
And in the end, who will really need to use such API? applications? no. Only one will take advantage of such API: the connection manager. Thus the conclusion: if the only user is application X, then do not make it as a public API, put it in x.

If you want an easier way to manage connectivity then it's in conn_mgr, you'll definitely benefit from having a primary generic interface that specific iface type can inherit and improve. See what NetworkManager or ConnMan are doing, it's really NIH.

But so far, you do not need this logic/design in the core stack.


Last but not least - though I really dislike using such authority argument - @jukkar and I have years of experience in connectivity management, and so far the various OSes, the various connection management services etc... always end up split like this for a reason: there is the logic that a network stack needs to handle and there is another one for the connection management service, and so far there has never been any point in pushing back to the stack some logic that is only meant to be used by the connection management service.
That does not prevent anybody to try doing things differently, but it has to come with really good use-cases/reasons.

glarsennordic commented:
> Thread joins, WiFi associates, ethernet does nothing, bluetooth does its own thing...
> And naming is extremely important, it brings semantics with it. If you put "connect" in l2 layer, so what is connect in ip layer then? It's messing up things. So then you get a connected iface, but it's not because it does not have any IP yet ? etc...

Actually, I think you're right. How about "attach/detach" instead, as a generalization of "associate/join"? This avoids the problem of a double-meaning for "connect".

> And in the end, who will really need to use such API? applications? no.

One use case we anticipate would be an application seeking to use only one network type, determined at build time, but buildable for any network type without changes to its network management logic.

> See what NetworkManager or ConnMan are doing, it's really NIH.

Can you point out the major difference you see between this PR and NetworkManager? The way I see it:
- NetworkManager defines a generic implementable interface for controlling attach/detach on network devices, as does this PR
- NetworkManager defines a generic implementable interface for configuring the settings used during attach/detach on network devices, as does this PR
- NetworkManager offers logic to maintain a connection on all available network devices, as does this PR

The only major difference, AFAICT, is that NetworkManager does this at the application level using pluggable DBus implementations, whereas we integrate at the net_if level.
We chose to implement at the net_if level, because attachment is necessary for an iface to become oper-up, and we felt it made sense to integrate a generic means of accessing this step at that level.

> To me you are trying to generalize concepts that cannot really be. And so far, the current stack already provide the means for doing all of this, per-interface type basis. Which all the stack need to provide.

Doesn't NetworkManager prove that the concepts we are trying to generalize have already been generalized by others successfully?

> Thus the conclusion: if the only user is application X, then do not make it as a public API, put it in x.

If the above does not convince you, then I suppose we could remove the modifications to `net_if` and make them live under `conn_mgr`.

The downside of this is that we would lose the ability to use the generic attach/detach calls without also including conn_mgr, which is something we specifically tried to avoid (see the use-case above).

Would you find this compromise (along with the rename to "attach/detach") satisfactory @tbursztyka ?

wbober commented:
> Last but not least - though I really dislike using such authority argument - @jukkar and I have years of experience in connectivity management, and so far the various OSes, the various connection management services etc... always end up split like this for a reason: there is the logic that a network stack needs to handle and there is another one for the connection management service, and so far there has never been any point in pushing back to the stack some logic that is only meant to be used by the connection management service. That does not prevent anybody to try doing things differently, but it has to come with really good use-cases/reasons.

@tbursztyka We appreciate the vast experience you, @jukkar, and other members of the community have. This is why we have reached out with an [RFC in Nov last year](https://github.com/zephyrproject-rtos/zephyr/issues/52695). The topic was also presented and then subsequently mentioned on 2 networking forums.

Would you be able to present, within a reasonable time, a more detailed design proposal - similar to what we have presented in the RFC? This would help us better understand how you imagine the solution and how decide for the next steps.
